### PR TITLE
gr.load_chat: Allow loading any openai-compatible server immediately as a ChatInterface

### PR DESCRIPTION
**Pull request by aliabid94**
gr.load_chat allows loading any openai chatbot immediately as a ChatInterface easily.

```python
gr.load_chat("http://localhost:11434/v1/", "llama3.2").launch()
```

Can be customized as such:
```python
import gradio as gr
gr.ChatInterface(
  gr.load_chat("http://localhost:11434/v1/", "llama3.2"),
  title="Etc"
).launch()
```

I think this will be an especially popular pairing with Ollama.

In a follow up PR, will add support for media input and responses.

### CODE REVIEW

**Comment by abidlabs** [comment]
Very cool @aliabid94! I find this part kind of confusing though:

> Can be customized as such:
> 
> ```py
> import gradio as gr
> gr.ChatInterface(
>   gr.load_chat("http://localhost:11434/v1/", "llama3.2"),
>   title="Etc"
> ).launch()
> ```

So `gr.load_chat` produces a ChatInterface, but now you can pass a ChatInterface as the `fn` parameter into a new ChatInterface? That's not intuitive and the DX is not optimal since many of the parameters in `gr.ChatInterface` (e.g. `additional_inputs`) don't make sense if you are loading an openai server. 

What about if we did the following instead?

`gr.ChatInterface.load(server_url)`. We could add `title` and whichever parameters actually make sense for loading an openai server to this factory method `load()` instead.

**Comment by aliabid94** [comment]
> What about if we did the following instead?
>
> gr.ChatInterface.load(server_url). To let users customize, we could add title and whichever parameters actually make sense for loading an openai server to this factory method gr.ChatInterface.load() instead.

Well we can't use `load`, that's the event listener for when the demo loads. So we could do something like `gr.ChatInterface.load_from(server_url)`, but that feels too long. I also don't really like the idea of copying all the kwargs and docstrings over again from the constructor to this load_from method, that would be like 40 lines of code duplication.

gr.load_chat mirrors gr.load. Using gr.load as a function for a customized Interface is also the current way to customize a gr.Interface, e.g.

```python
demo = gr.Interface(fn=gr.load("model/meta-ai/llama3.2"), title="My App", examples=["a", "b"])
```

mirrors

```python
demo = gr.ChatInterface(fn=gr.load_chat("http://chat.com/v1"), title="My App", examples=["a", "b"])
```

So I think the existing API makes the most sense.

**Comment by abidlabs** [comment]
I see thanks for explaining 👍

File: gradio/external.py [hunk 1]
```diff
@@ -581,3 +582,75 @@ def fn(*data):
...
+        response = ""
+        for chunk in stream:
+            response += chunk.choices[0].delta.content
+            yield response
+
+    return ChatInterface(open_api_stream if streaming else open_api, type="messages")
+
+
+def format_tuples_if_needed(history: list) -> list[dict]:
+    if len(history) == 0 or not isinstance(history[0], list):
```
Comment (abidlabs) [inline gradio/external.py:648]:
```suggestion
    if len(history) == 0 or not isinstance(history[0], (list, tuple)):
```

File: gradio/external.py [hunk 2]
```diff
@@ -581,3 +582,75 @@ def fn(*data):
...
+    from gradio.chat_interface import ChatInterface
+
+    client = OpenAI(api_key=token, base_url=base_url)
+    start_message = (
+        [{"role": "system", "content": system_message}] if system_message else []
+    )
+
+    def open_api(message: str, history: list | None) -> str:
+        history = history or start_message
+        history = format_tuples_if_needed(history)
```
Comment (abidlabs) [inline gradio/external.py:619]: Just to confirm, the only reason you need this is someone decides they want to use `gr.load_chat` and then pass in the resulting `gr.ChatInterface` into another one in which `type="tuples`"?

File: gradio/external.py [hunk 3]
```diff
@@ -581,3 +582,75 @@ def fn(*data):
...
+            yield response
+
+    return ChatInterface(open_api_stream if streaming else open_api, type="messages")
+
+
+def format_tuples_if_needed(history: list) -> list[dict]:
+    if len(history) == 0 or not isinstance(history[0], list):
+        return history
+    messages = []
+    for user, assistant in history:
```
Comment (abidlabs) [inline gradio/external.py:651]: Can you use `gr.ChatInterface._tuples_to_messages` instead of duplicating?

**Review by abidlabs — APPROVED** [review]
lgtm works great @aliabid94! Can we add some documentation around this at the top of https://www.gradio.app/guides/creating-a-chatbot-fast?

File: gradio/external.py [hunk 4]
```diff
@@ -581,3 +582,75 @@ def fn(*data):
     kwargs["_api_mode"] = True
     interface = gradio.Interface(**kwargs)
     return interface
+
+
+def load_chat(
```
Comment (abidlabs) [inline gradio/external.py:587]: Let's @document this

File: gradio/external.py (hunk 2 above)
Comment (aliabid94) [reply gradio/external.py:619]: yep exactly

File: guides/05_chatbots/01_creating-a-chatbot-fast.md [hunk 5]
```diff
@@ -14,6 +14,16 @@ This tutorial uses `gr.ChatInterface()`, which is a high-level abstraction that
...
 ```
 
+## Quickly loading from Ollama or any "OpenAI API compatible" endpoint
+
+If you have a chat server serving an OpenAI API compatible endpoint (skip ahead if you don't), you can spin up a ChatInterface in a single line. First, also run `pip install openai`. Then, with your own URL, model, and optional token:
+
+```python
+import gradio as gr
+
+gr.load_chat("http://localhost:11434/v1/", "llama3.2").launch()
```
Comment (abidlabs) [inline guides/05_chatbots/01_creating-a-chatbot-fast.md:24]:
```suggestion
gr.load_chat("http://localhost:11434/v1/", model="llama3.2", token="ollama").launch()
```

File: guides/05_chatbots/01_creating-a-chatbot-fast.md [hunk 6]
```diff
@@ -14,6 +14,16 @@ This tutorial uses `gr.ChatInterface()`, which is a high-level abstraction that
 $ pip install --upgrade gradio
 ```
 
+## Quickly loading from Ollama or any "OpenAI API compatible" endpoint
```
Comment (abidlabs) [inline guides/05_chatbots/01_creating-a-chatbot-fast.md:17]:
```suggestion
## Quickly loading from Ollama or any OpenAI-API compatible endpoint
```

**Pull request merged**